drivers/SmartThings/zigbee-switch/src/aqara/init.lua

```diff
@@ -70,9 +112,36 @@ local function is_aqara_products(opts, driver, device)
   return false
 end
 
-local function emit_power_consumption_report_event(device, value)
-  local raw_value = value.value -- 'Wh'
+local function private_mode_handler(driver, device, value, zb_rx)
+  device:set_field(PRIVATE_MODE, value.value, { persist = true })
+end
+
+local function on_off_handler(driver, device, value, zb_rx)
+  device:emit_event_for_endpoint(
+    zb_rx.address_header.src_endpoint.value,
+    value.value and capabilities.switch.switch.on() or capabilities.switch.switch.off()
+  )
+
+  local private_mode = device:get_field(PRIVATE_MODE) or 0
+  if private_mode == 1 then
+    -- read power meter
+    device:send(AnalogInput.attributes.PresentValue:read(device):to_endpoint(POWER_METER_ENDPOINT))
+  end
+end
+
+local function wireless_switch_handler(driver, device, value, zb_rx)
+  local endpoint = wireless_switch_endpoint_map[zb_rx.address_header.src_endpoint.value]
+  local event = wireless_switch_button_event_map[value.value]
+  if event ~= nil then
+    device:emit_event_for_endpoint(endpoint, event)
```

Comment:
So the child devices emit events from _both_ endpoints `1,2,3` but also `0x29, 0x2A, 0x2B`? You could just change your find_child function to subtract 40 for endpoint sources over 3 and then avoid this extra mapping work.

Reply:
I didn't understand what you said. Can you tell me more specific?
When the user presses a switch (button) of the device, a switch press event is received. Users can create automations using these events as triggers.
The log is below.

Button 1:

received Zigbee message: 
< ZigbeeMessageRx || type: 0x00, < AddressHeader || src_addr: 0x999F, src_endpoint: 0x2A, dest_addr: 0x0000, dest_endpoint: 0x01, profile: 0x0104, cluster: 0x0012 >, lqi: 0xC8, rssi: -80, body_length: 0x0008, < ZCLMessageBody || < ZCLHeader || frame_ctrl: 0x18, seqno: 0x08, ZCLCommandId: 0x0A >, < ReportAttribute || < AttributeRecord || AttributeId: 0x0055, DataType: Uint16, Uint16: 0x0001 > > > >

emitting event: 
{"attribute_id":"button","capability_id":"button","component_id":"main","state":{"value":"pushed"},"state_change":true}
2023-02-22T03:42:52.761736341+00:00 DEBUG Zigbee Switch  Aqara Smart Wall Switch (No Neutral, Double Rocker) 1 device thread event handled

Button 2:

received Zigbee message: 
< ZigbeeMessageRx || type: 0x00, < AddressHeader || src_addr: 0x999F, src_endpoint: 0x29, dest_addr: 0x0000, dest_endpoint: 0x01, profile: 0x0104, cluster: 0x0012 >, lqi: 0xCA, rssi: -64, body_length: 0x0008, < ZCLMessageBody || < ZCLHeader || frame_ctrl: 0x18, seqno: 0x0A, ZCLCommandId: 0x0A >, < ReportAttribute || < AttributeRecord || AttributeId: 0x0055, DataType: Uint16, Uint16: 0x0001 > > > >

emitting event: 
{"attribute_id":"button","capability_id":"button","component_id":"main","state":{"value":"pushed"},"state_change":true}

Reply:
The `emit_event_for_endpoint` function first tries to use the device's defined `find_child` function, and then falls back to the `get_component_for_endpoint` function (for multi-component devices). You have your regular mapping of children, which seems to be `endpoint 1 -> switch 1`, `endpoint 2 -> switch 2`, etc. But you're also mapping these events from other endpoints to these same switch children. If you changed your `find_child` function to also map events from `endpoint 0x29 -> switch 1`, you wouldn't have to do this separate lookup here, you could just call `emit_event_for_endpoint([button event], 0x29)` and have it automatically be emitted by switch 1.

Reply:
Good idea 👍 I updated code. Please review it. Thanks.